scripts/evaluation_utils.py

```diff
@@ -0,0 +1,162 @@
+import argparse
+
+import numpy as np
+
+import torch
+from torch import nn
+from torch.utils.data import Dataset
+
+from tqdm import tqdm
+
+
+def configure_arg_parser():
+    parser = argparse.ArgumentParser()
+
+    parser.add_argument("-m", "--model", help="Seq2seq model name from huggingface")
+    parser.add_argument("-p", "--path", help="Path to dataset for evaluation")
+    parser.add_argument("-bs", "--batch_size", type=int, default=64, help="Batch size for evaluation")
+    parser.add_argument("-s", "--side", default="left", help="Side of truncation and padding for tokenizer")
+    parser.add_argument("-cl", "--max_context", type=int, default=256, help="Max length of context fed into model")
+    parser.add_argument("-gl", "--max_gen", type=int, default=64, help="Max length of generated output")
+    parser.add_argument("-d", "--device", default="cpu", help="Device on which to evaluate the model")
+    parser.add_argument("-ds", "--do_sample", default=True, help="parameter for generation method")
+    parser.add_argument("-nb", "--num_beams", default=4, help="Parameter for generation method")
+
+    return parser
+
+
+class ConvAI2Dataset(Dataset):
+    """ConvAI2 Dataset"""
+
+    def __init__(self, path):
+        """
+        Args:
+            path (str): File path to ConvAI2 dataset
+        """
+
+        persona = "your persona"
+        context = ""
+
+        candidates = []
+
+        dataset = {"context": [], "candidates": []}
+
+        with open(path, "r") as f:
+            for line in f.readlines():
+                if line[2 : 2 + len(persona)] == persona:
+                    context = ""
+                    candidates = []
+                    continue
+
+                else:
+                    prev_candidates = candidates
+
+                    utterance1, utterance2, _, candidates = line[2:].split("\t")
+                    candidates = candidates.split("|")[:-1]
+
+                    if prev_candidates:
+                        dataset["context"].append(context)
+                        dataset["candidates"].append([utterance1] + candidates)
+
+                    context += "<s>" + utterance1 + "</s>" + "<s>" + utterance2 + "</s>"
+
+        self.dataset = dataset
+
+    def __len__(self):
+        return len(self.dataset["context"])
+
+    def __getitem__(self, idx):
+        sample = {
+            "context": self.dataset["context"][idx],
+            "candidates": self.dataset["candidates"][idx],
+        }
+
+        return sample
+
+
+def calculate_hits_ppl(model, tokenizer, dataset, max_len, device):
+    hits = []
+    ppl = []
+
+    loss_fn = nn.CrossEntropyLoss(reduction="none")
```

Comment:
I think that there is no need for separate class, you can use [`torch.nn.functional.cross_entropy`](https://pytorch.org/docs/stable/generated/torch.nn.functional.cross_entropy.html#torch.nn.functional.cross_entropy)